python/ql/src/semmle/python/objects/Callables.qll

```diff
@@ -0,0 +1,460 @@
+        origin = CfgOrigin::fromCfgNode(this.getScope().getEntryNode())
+    }
+
+    override predicate calleeAndOffset(Function scope, int paramOffset) {
+        scope = this.getScope() and paramOffset = 0
+    }
+
+    override string getName() {
+        result = this.getScope().getName()
+    }
+
+    override boolean isDescriptor() { result = true }
+
+    pragma [noinline] override predicate descriptorGetClass(ObjectInternal cls, ObjectInternal value, CfgOrigin origin) {
+        any(ObjectInternal obj).binds(cls, _, this) and
+        value = this and origin = CfgOrigin::fromCfgNode(this.getOrigin())
+    }
+
+    pragma [noinline] override predicate descriptorGetInstance(ObjectInternal instance, ObjectInternal value, CfgOrigin origin) {
+        value = TBoundMethod(instance, this) and origin = CfgOrigin::unknown()
+    }
+
+    override CallNode getACall(PointsToContext ctx) {
+        PointsTo::pointsTo(result.getFunction(), ctx, this, _)
+        or
+        exists(BoundMethodObjectInternal bm |
+            bm.getACall(ctx) = result and this = bm.getFunction()
+        )
+    }
+
+    override NameNode getParameter(int n) {
+        result.getNode() = this.getScope().getArg(n)
+    }
+
+    override NameNode getParameterByName(string name) {
+        result.getNode() = this.getScope().getArgByName(name)
+    }
+
+
+    override predicate neverReturns() {
+        InterProceduralPointsTo::neverReturns(this.getScope())
+    }
+
+    override predicate functionAndOffset(CallableObjectInternal function, int offset) {
+        function = this and offset = 0
+    }
+
+}
+
+
+private BasicBlock blockReturningNone(Function func) {
+    exists(Return ret |
+        not exists(ret.getValue()) and
+        ret.getScope() = func and
+        result = ret.getAFlowNode().getBasicBlock()
+    )
+}
+
+
+/** Class representing built-in functions such as `len` or `print`. */
+class BuiltinFunctionObjectInternal extends CallableObjectInternal, TBuiltinFunctionObject {
+
+    override Builtin getBuiltin() {
+        this = TBuiltinFunctionObject(result)
+    }
+
+    override string toString() {
+        result = "Builtin-function " + this.getBuiltin().getName()
+    }
+
+    override predicate introducedAt(ControlFlowNode node, PointsToContext context) {
+        none()
+    }
+
+    override ObjectInternal getClass() {
+        result = TBuiltinClassObject(this.getBuiltin().getClass())
+    }
+
+    override boolean testableForEquality() { result = true }
+
+    override predicate callResult(PointsToContext callee, ObjectInternal obj, CfgOrigin origin) { none() }
+
+    pragma [noinline]
+    override predicate callResult(ObjectInternal obj, CfgOrigin origin) {
+        exists(Builtin func, BuiltinClassObjectInternal cls |
+            func = this.getBuiltin() and
+            func != Builtin::builtin("isinstance") and
+            func != Builtin::builtin("issubclass") and
+            func != Builtin::builtin("callable") and
```

Comment:
They are handled specially in PointsTo.qll as their values depend on the class of their argument.
They could be handled by `callResult/3`, but because they are closely related to the class hierarchy, it makes more sense to leave them in PointsTo.qll.